### DISCUSSION TITLE
Recipe for simple `select some_function()` statement without a table

### DISCUSSION THREAD

**Discussion by tgross35** [diesel-rs/diesel#3407] [Q&A] [answered]
Hello everyone,

What is the most simple way to run something like `select some_function()` that is independent of a table?

I am looking for something like the following that returns a nullable integer, but have run into issues with `load_dsl::private::CompatibleType` being implemented:

```rust
sql_query("select some_function()").load::<(Option<i64>,)>(conn);
```

It seems like it should be pretty simple, but I am definitely missing something.

**Comment by tgross35** [comment]
Ok - I have something that works!

```rust
let res: (i32,) = sql::<(Integer,)>("select udf_sequence()")
    .get_result(conn)
    .expect("bad result");
```

My best interpretation of the above is:
- `sql` needs a generic identifying the SQL type (`(Integer,)` here)
- `get_result` returns a `Result<something, diesel::result::Error>`
- `get_result` tries to coerce the generic SQL type into `something`, whatever that is. I think these will be Rust types, in this case an `(i32,)` tuple

It could be good to have a `core` section of the website for ORM-less stuff, kind of like `sqlalchemy` has - this stuff is tricky!

**Reply by tgross35** [reply]
Adding to the examples, this works to use `execute` to run something that doesn't return

```rust
sql::<(Untyped,)>(stmt)
    .execute(conn)
    .expect("could not run setup");
```

At a minimum I think it would be good to throw these examples in the docs. If you would like me to submit a PR myself I will; however, I feel like there is a lot of context that I personally don't know about that would be good to include

**Reply by weiznich** [reply]
I'm always happy for documentation improvements here, but I cannot see how that would be easily possible. The [crate documentation](https://docs.diesel.rs/master/diesel/index.html#serializing-and-deserializing) already has plenty of details in [multiple](https://docs.diesel.rs/master/diesel/prelude/trait.Queryable.html) [locations](https://docs.diesel.rs/master/diesel/deserialize/trait.QueryableByName.html) about how deserialization works and people keep missing that.

**Reply by tgross35** [reply]
I guess I meant less documentation and more examples. I couldn’t find any examples for doing what I wanted to do on the website or the rustdoc - every example seems to be for ORM, returning a struct.

Additionally confusing for me is that the examples have a lot of hidden code that gives background on where the type coercion comes from. But that’s all visible in the source of course

**Reply by weiznich** [reply]
> I guess I meant less documentation and more examples. I couldn’t find any examples for doing what I wanted to do on the website or the rustdoc - every example seems to be for ORM, returning a struct.

As already written: We are happy to accept improvements to the documentation. That said: We probably will never cover all use cases with an example, as in the end SQL it a quite complex language and you can combine almost anything with anything else, which results in an combinatoric explosion in examples if you try to cover everything. That's the reason why most of the examples in the API documentation focus on one feature. For more complex interactions we have the guides on the web page. We are aware that there are at least two guides missing ("All about queries" and probably something about deserialization in detail). Writing these guides is quite a bit of work. We miss currently the capacity to write these guides. 

> Additionally confusing for me is that the examples have a lot of hidden code that gives background on where the type coercion comes from.

Including all code to make any example self-contained would result in quite long examples. That will result in people not finding the relevant example as well, if it's not the first one they look at. This is especially true for more complex patterns.

**Comment by weiznich** [comment, marked as answer]
The preferred way of writing such functionality is by using [ `diesel::dsl::select`](https://docs.diesel.rs/master/diesel/dsl/fn.select.html) + [`diesel::sql_function!`](https://docs.diesel.rs/master/diesel/prelude/macro.sql_function.html).

This allows to write code such as:

```rust
sql_function! {
    fn some_function() -> Nullable<Integer>;
}

diesel::select(some_function()).get_result::<Option<i32>>(conn);

```

`sql_query("…").get_result::<Option<i32>>(conn)` does not work as [this function is documented to require a result type that implements `QueryableByName`](https://docs.diesel.rs/master/diesel/fn.sql_query.html). It's even explicit about that this is only the case for types with an explicit `#[derive(QueryableByName)]` in your crate.

**Reply by tgross35** [reply]
That’s awesome! Thank you
